src/Character/Investigator.java

```diff
@@ -1,75 +1,260 @@
-	 * is that the exact role of his target is revealed.
-	 */
-	public Investigator(Player player) {
-		super(Roles.Investigator, player);
-		detectsRole = GameOptions.InvestigatorOptions.DetectRole.DETECT_VAGUE_ROLE;
-	}
+	private Random random_generator;
 
-	/**
-	 * Investigator. Game Option constructor.
-	 * 
-	 * @param gameOption
-	 *            - Investigator.Options.DETECT_EXACT_ROLE: the investigator
-	 *            will discover the exact role of his target -
-	 *            Investigator.Options.DETECT_VAGUE_ROLE: the investigator will
-	 *            discover a hint for the role of his target
-	 */
-	Investigator(Player player, GameOptions.Options detectsRole) {
+	protected Investigator(Player player) {
 		super(Roles.Investigator, player);
-		this.detectsRole = detectsRole;
+		random_generator = new Random();
 	}
 
 	@Override
-	public String doAction() throws CannotGetPlayerException {
-		
-		if (getTarget().size() != 1) {
-			return GameMessage.NO_ACTION;
+	public boolean performAction(List<Player> alive_player, Character target) {
+		if (this.character_status.isBlocked()) {
+			result = GameLog.Character.BLOCKED(player);
+			return true;
 		}
-		
-		Player target = getTarget().get(0);
-		
-		if(!GameEngine.getCharacter(target).isAlive()) {
-			return GameMessage.TARGET_DEAD;
+		target.character_status.getVisitors().add(player);
+		List<Investigations> investigation_list = Investigations
+				.doInvestigation(target.getRole());
+		int index = random_generator.nextInt(investigation_list.size());
+		result = GameLog.Character.INVESTIGATION_SUCCESS(player,
+				target.player, investigation_list.get(index).toString());
+		return true;
+	}
+
+	private enum Investigations {
+		Trespassing("Trespassing"), Kidnapping("Kidnapping"), NoCrime(
+				"No Crime"), Corruption("Corruption"), IdentityThief(
+				"Identity Thief"), Soliciting("Soliciting"), Murder("Murder"), DisturbanceOfPeace(
+				"Disturbance of Peace"), Conspiracy("Conspiracy"), DestructionOfProperty(
+				"Destruction of Property"), Arson("Arson");
+
+		String investigation;
+
+		Investigations(String investigation) {
+			this.investigation = investigation;
 		}
-		if (this.isRoleBlocked()) {
-			return GameMessage.ROLE_BLOCKED;
+
+		@Override
+		public String toString() {
+			return investigation.toLowerCase();
 		}
-		
-		GameEngine.getCharacter(target).addVisitor(getPlayer());
-		
-		DetectRole detect = (DetectRole) detectsRole;
-		
-		switch (detect) {
-		case DETECT_EXACT_ROLE:
-			return GameMessage.INVESTIGATION_EXACT(target.getName());
-		case DETECT_VAGUE_ROLE:
-			return GameMessage.INVESTIGATION_VAGUE(target.getName());
-		default:
-			return GameMessage.INVESTIGATION_VAGUE(target.getName());
+
+		/**
+		 * Makes a list of possible investigations for a role
+		 * 
+		 * @param role
+		 * @return list of possible investigation results for that role
+		 */
+		protected static List<Investigations> doInvestigation(Roles role) {
```

Comment:
Modularity is about improving readability and maintainability, but necessarily write something with as few lines as possible or reusing a piece of code.

In the case of only two classes, maintainability isn't that big of an issue especially with test cases, but adding abstract classes really intensifies the complexity for readability.

But that's just my opinion.

Performance in this case is also negligible.

I think Eleanor wanted a more elegant way of writing this method.

The best solution I can think of for this particular problem would probably be Spring's Framework for Inversion of Control. It works the same way as a character factory.

What it does it it has a XML file of 'beans'. The 'beans' are blueprints to to creating an object.

So in this case the whole switch statement can be simplified down to two lines by making the wrapper for the beanfactory, and then passing the role to the beanfactory and getting back the array list.

It could also do justice to move the enum to the parent class.